#!/bin/bash

# final genomic in-silico pcr
# the input file should have this format NAME\tFOR_SEQ\tREV_SEQ
${PRIMERSEARCH} \
-seqall ${GENOME} \
-infile val_cand_4genomcheck.txt \
-mismatchpercent ${PMISMATCH} \
-snucleotide1 \
-outfile final_gnm.primersearch

# count genomic aspecifics
# are they <= 3000bp and therefore real?
grep 'Amplimer length' *gnm.primersearch | cut -d' ' -f3 > genomic_amplimers.txt

# How many amplicons under 10kb are generated by perfect match with the genome?

THE='final_gnm.primersearch'
OOT='genomic_mismatch.txt'
touch $OOT

AMP=($(grep -E -n "length: [0-9]{3,4} bp" $THE | cut -d: -f1))
PRM=($(grep -n "Primer" $THE | cut -d: -f1))

for e in ${AMP[*]}; do
	sed -n "$(($e-5)),$e"p $THE > tempura.temp
	[ "$(grep -E "[0-4] mismatches" tempura.temp | wc -l)" -eq 2 ] && \
	cat tempura.temp >> $OOT && \
	for j in ${PRM[*]}; do
		[ "$j" -gt "$e" ] && sed -n "$(($j - 1))"p $THE >> $OOT && break;
	done;
done
rm tempura.temp

# genomic_mismatch.txt should be EMPTY
# if the primers designed are specific
